File: solutions/javascript/graphs/bfs/constructShortestPath.js
const NO_VALID_PATH = [];

class Queue {
    constructor(array = []) {
        this.elements = array;

        this.startIdx = 0;
        this.endIdx = array.length - 1;
    }

    enqueue(newElement) {
        this.elements.push(newElement);
        this.endIdx++;
    }

    dequeue() {
        if (this.isEmpty()) return null;

        const returnElement = this.elements[this.startIdx];

        this.startIdx++;
        return returnElement;
    }

    isEmpty() {
        return this.size() <= 0;
    }

    size() {
        return this.endIdx - this.startIdx + 1;
    }
}

const buildGraph = (edges) => {
    const graph = {};

    for (const [nodeOne, nodeTwo] of edges) {
        const nodeOneInGraph = graph.hasOwnProperty(nodeOne);
        const nodeTwoInGraph = graph.hasOwnProperty(nodeTwo);

        if (!nodeOneInGraph) graph[nodeOne] = [];
        if (!nodeTwoInGraph) graph[nodeTwo] = [];

        graph[nodeOne].push(nodeTwo);
        graph[nodeTwo].push(nodeOne);
    }

    return graph;
};

const constructPath = (startCity, endCity, idealPrevCityMap) => {
    let curCity = endCity;
    const path = [];

    while (curCity !== null) {
        path.push(curCity);
        curCity = idealPrevCityMap[curCity];
    }

    return path.reverse();
};

const getShortestPath = (startCity, endCity, roads) => {
    const graph = buildGraph(roads);

    const queue = new Queue();
    queue.enqueue({
        city: startCity,
        distanceFromStart: 0,
        prevCity: null,
    });

    const visited = new Set();
    const idealPrevCityMap = {};

    while (queue.size() > 0) {
        // Remove node
        const { city, distanceFromStart, prevCity } = queue.dequeue();

        // Process node
        visited.add(city);

        const cityInIdealPrevCityMap = idealPrevCityMap.hasOwnProperty(city);

        if (!cityInIdealPrevCityMap) idealPrevCityMap[city] = prevCity;
        if (city === endCity) break;

        // Add neighbors
        const neighbors = graph[city];
        for (const neighborCity of neighbors) {
            if (visited.has(neighborCity)) continue;

            queue.enqueue({
                city: neighborCity,
                distanceFromStart: distanceFromStart + 1,
                prevCity: city,
            });
        }
    }

    const endCityInIdealPrevCityMap = idealPrevCityMap.hasOwnProperty(endCity);
    if (!endCityInIdealPrevCityMap) return NO_VALID_PATH;

    return constructPath(startCity, endCity, idealPrevCityMap);
};

const startCity = 5;
const endCity = 10;
const roads = [
    [5, 7],
    [5, 3],
    [7, 6],
    [7, 4],
    [3, 9],
    [6, 4],
    [4, 10],
    [4, 9],
];

console.log(getShortestPath(startCity, endCity, roads));

File: solutions/python/bfs/construct_shortest_path.py
"""
Given 3 inputs: a start city, an end city, and a list of roads, return the shortest path (as an array) from the start city to the end city.
"""

from collections import deque, defaultdict

no_valid_path = []

class CityNode:
    def __init__(self, city, distance_from_start, prev_city):
        self.city = city
        self.distance_from_start = distance_from_start
        self.prev_city = prev_city

def build_graph(edges):
    graph = defaultdict(list)

    for node_one, node_two in edges:
        graph[node_one].append(node_two)
        graph[node_two].append(node_one)

    return graph

def construct_path(start_city, end_city, ideal_prev_city_map):
    cur_city = end_city
    path = []

    while cur_city is not None:
        path.append(cur_city)
        cur_city = ideal_prev_city_map[cur_city]
    
    path.reverse()
    return path

def get_shortest_path(start_city, end_city, roads):
    graph = build_graph(roads)

    queue = deque()
    start_node = CityNode(start_city, 0, None)
    queue.append(start_node)

    visited = set()
    ideal_prev_city_map = {}

    while queue:
        # Remove node
        city_node = queue.popleft()

        city = city_node.city
        distance_from_start = city_node.distance_from_start
        prev_city = city_node.prev_city

        # Process node
        visited.add(city)

        if city not in ideal_prev_city_map:
            ideal_prev_city_map[city] = prev_city

        if city == end_city:
            break

        # Add neighbors
        neighbors = graph[city]
        for neighbor_city in neighbors:
            if neighbor_city in visited:
                continue
        
            neighbor_node = CityNode(neighbor_city, distance_from_start + 1, city)
            queue.append(neighbor_node)

    if end_city not in ideal_prev_city_map:
        return no_valid_path

    return construct_path(start_city, end_city, ideal_prev_city_map)

start_city = 5
end_city = 10
roads = [[5,7], [5,3], [7,6], [7,4], [3,9], [6,4], [4,10], [4,9]]

print(f"Your answer: {get_shortest_path(start_city, end_city, roads)}")
print(f"Correct answer: {[5,7,4,10]}")
print()





